File: 백준/Gold/2143. 두 배열의 합/두 배열의 합.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;

    public static void main(String[] args) throws IOException {
        long T = Long.parseLong(br.readLine());

        int n = Integer.parseInt(br.readLine());
        int[] n_num = new int[n];
        List<Long> n_part = new ArrayList<>();

        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            n_num[i] = Integer.parseInt(st.nextToken());
        }

        for (int i = 0; i < n; i++) {
            n_part.add((long) n_num[i]);
            long total = n_num[i];
            for (int j = i + 1; j < n; j++) {
                total += n_num[j];
                n_part.add(total);
            }
        }

        int m = Integer.parseInt(br.readLine());
        Map<Long, Integer> mapM = new HashMap<>();
        int[] m_num = new int[m];

        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < m; i++) {
            m_num[i] = Integer.parseInt(st.nextToken());
        }

        for (int i = 0; i < m; i++) {
            long total = m_num[i];
            if (mapM.containsKey(total)) {
                mapM.put(total, mapM.get(total) + 1);
            } else {
                mapM.put(total, 1);
            }

            for (int j = i + 1; j < m; j++) {
                total += m_num[j];
                if (mapM.containsKey(total)) {
                    mapM.put(total, mapM.get(total) + 1);
                } else {
                    mapM.put(total, 1);
                }
            }
        }

        long cnt = 0L;
        for (int i = 0; i < n_part.size(); i++) {
            long check = T - n_part.get(i);
            if (mapM.containsKey(check)) {
                cnt += mapM.get(check);
            }
        }
        System.out.println(cnt);
    }
}

File: 백준/Gold/2143. 두 배열의 합/두 배열의 합.py
import sys
input = sys.stdin.readline

t = int(input())

n = int(input())
n_num = list(map(int,input().split()))
n_part = []

m = int(input())
m_num = list(map(int,input().split()))
m_part = []

for i in range(n):
    n_part.append(n_num[i])
    total = n_num[i]
    for j in range(i+1, n):
        total += n_num[j]
        n_part.append(total)
n_part.sort()

dic = {}

for i in range(m):
    m_part.append(m_num[i])
    total = m_num[i]
    if m_num[i] in dic:
        dic[m_num[i]] += 1
    else:
        dic[m_num[i]] = 1
        
    for j in range(i+1, m):
        total += m_num[j]
        if total in dic:
            dic[total] += 1
        else:
            dic[total] = 1
        m_part.append(total)

m_part.sort()

cnt= 0
for i in n_part:
    if t - i in dic:
        cnt += dic[t-i]

print(cnt)
